{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "17584646-7428-485c-a801-aea09a8d0fa9",
   "metadata": {},
   "source": [
    "# pyUCell - some important parameters"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7022dae5-2491-49a9-8486-cd0a1277f3ea",
   "metadata": {},
   "source": [
    "This document describes some important parameters of the pyUCell algorithm, and how they can be adapted depending on your dataset.\n",
    "\n",
    "For a simple pyUCell tutorial refer to: [pyUCell basics](https://pyucell.readthedocs.io/en/latest/notebooks/basic.html#)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c633515-51b5-481a-a9f1-cd9044d53028",
   "metadata": {},
   "source": [
    "## Load example dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a6715005-6ca7-4a1a-a0db-873ad888e3fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pyucell as uc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "338c4bd3-093c-475b-8f09-b101ab7b6012",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = sc.datasets.pbmc3k()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4387cc5-0081-40fb-85a2-38eee84ae393",
   "metadata": {},
   "source": [
    "**Note:** becase UCell scores are based on relative gene ranks, it can be applied both on raw counts or normalized data. As long as the normalization preserves the relative ranks between genes, the results will be equivalent."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b843524a-386d-4966-868f-c95ce29559e1",
   "metadata": {},
   "source": [
    "## 1. Positive and negative gene sets in signatures"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9400ca05-7830-4f38-a007-bcd03f4e9f1b",
   "metadata": {},
   "source": [
    "pyUCell supports **positive and negative gene sets** within a signature. Simply append + or - signs to the genes to include them in positive and negative sets, respectively. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "17f9ec9e-6df5-4e5a-aa65-95a853e748e7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "signatures = {\"CD8T\": [\"CD8A+\", \"CD8B+\", \"CD4-\"], \"CD4T\": [\"CD4+\", \"CD40LG+\", \"CD8A-\", \"CD8B-\"]}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3d99f25-651f-4094-abae-8f9c63c28d11",
   "metadata": {},
   "source": [
    "pyUCell evaluates the positive and negative gene sets separately, then subtracts the scores. The parameter `w_neg` controls the relative weight of the negative gene set compared to the positive set (`w_neg=1.0` means equal weight). Note that the combined score is clipped to zero, to preserve UCell scores in the [0, 1] range."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "83cf601f-d091-4da1-9ac2-2b7ec7ca3da2",
   "metadata": {},
   "outputs": [],
   "source": [
    "uc.compute_ucell_scores(adata, signatures=signatures, w_neg=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a712b0de-8882-4750-b88c-8b13cce0ab48",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD8T_UCell</th>\n",
       "      <th>CD4T_UCell</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACAACCAC-1</th>\n",
       "      <td>0.770771</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGAGCTAC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGATCAGC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.252586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGCTTCCG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGTATGCG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCGAACTCTCAT-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCTACTGAGGCA-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCTACTTCCTCG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGAGAGGC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGCCTCAC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2700 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CD8T_UCell  CD4T_UCell\n",
       "index                                   \n",
       "AAACATACAACCAC-1    0.770771    0.000000\n",
       "AAACATTGAGCTAC-1    0.000000    0.000000\n",
       "AAACATTGATCAGC-1    0.000000    0.252586\n",
       "AAACCGTGCTTCCG-1    0.000000    0.000000\n",
       "AAACCGTGTATGCG-1    0.000000    0.000000\n",
       "...                      ...         ...\n",
       "TTTCGAACTCTCAT-1    0.000000    0.000000\n",
       "TTTCTACTGAGGCA-1    0.000000    0.000000\n",
       "TTTCTACTTCCTCG-1    0.000000    0.000000\n",
       "TTTGCATGAGAGGC-1    0.000000    0.000000\n",
       "TTTGCATGCCTCAC-1    0.000000    0.000000\n",
       "\n",
       "[2700 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ae7b062-da8c-4213-a26c-f84ca89a9328",
   "metadata": {},
   "source": [
    "## 2. The `max_rank` parameter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5a4ef28-1f26-4f3c-b7ce-a4ed0e415c4d",
   "metadata": {},
   "source": [
    "Single-cell data are sparse. In other words, for any given cell only a few hundred/a few thousand genes (out of tens of thousands) are detected with at least one UMI count. Because UCell scores are based on ranking genes by their expression values, it is essential to account for data sparsity when calculating ranks. This is implemented by capping ranks to a `max_rank` parameter, in other words only the top `max_rank` genes are ranked, and the rest are assumed equivalent at the lowest ranking value.\n",
    "\n",
    "It is often useful to adjust the `max_rank` depending on the sparsity of your dataset. A good rule of thumb is to examine the median number of expressed genes per cell, and set `max_rank` in that order of magnitude. For example, for the test dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "8d173d8e-1884-43b8-8d1a-6f13b45bf706",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'function' object has no attribute 'X'",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mAttributeError\u001b[39m                            Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[27]\u001b[39m\u001b[32m, line 2\u001b[39m\n\u001b[32m      1\u001b[39m aa = adata.copy\n\u001b[32m----> \u001b[39m\u001b[32m2\u001b[39m \u001b[43msc\u001b[49m\u001b[43m.\u001b[49m\u001b[43mpp\u001b[49m\u001b[43m.\u001b[49m\u001b[43mcalculate_qc_metrics\u001b[49m\u001b[43m(\u001b[49m\u001b[43maa\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43minplace\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[32m      4\u001b[39m sc.pl.violin(aa, \u001b[33m\"\u001b[39m\u001b[33mn_genes_by_counts\u001b[39m\u001b[33m\"\u001b[39m, jitter=\u001b[32m0.4\u001b[39m, size=\u001b[32m1\u001b[39m, log=\u001b[38;5;28;01mTrue\u001b[39;00m, groupby=\u001b[38;5;28;01mNone\u001b[39;00m)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.11.11/envs/scverse/lib/python3.11/site-packages/scanpy/preprocessing/_qc.py:301\u001b[39m, in \u001b[36mcalculate_qc_metrics\u001b[39m\u001b[34m(adata, expr_type, var_type, qc_vars, percent_top, layer, use_raw, inplace, log1p, parallel)\u001b[39m\n\u001b[32m    296\u001b[39m     warn(\n\u001b[32m    297\u001b[39m         \u001b[33m\"\u001b[39m\u001b[33mArgument `parallel` is deprecated, and currently has no effect.\u001b[39m\u001b[33m\"\u001b[39m,\n\u001b[32m    298\u001b[39m         \u001b[38;5;167;01mFutureWarning\u001b[39;00m,\n\u001b[32m    299\u001b[39m     )\n\u001b[32m    300\u001b[39m \u001b[38;5;66;03m# Pass X so I only have to do it once\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m301\u001b[39m X = \u001b[43m_choose_mtx_rep\u001b[49m\u001b[43m(\u001b[49m\u001b[43madata\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43muse_raw\u001b[49m\u001b[43m=\u001b[49m\u001b[43muse_raw\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlayer\u001b[49m\u001b[43m=\u001b[49m\u001b[43mlayer\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    302\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(X, coo_matrix):\n\u001b[32m    303\u001b[39m     X = csr_matrix(X)  \u001b[38;5;66;03m# COO not subscriptable\u001b[39;00m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~/.pyenv/versions/3.11.11/envs/scverse/lib/python3.11/site-packages/scanpy/preprocessing/_qc.py:45\u001b[39m, in \u001b[36m_choose_mtx_rep\u001b[39m\u001b[34m(adata, use_raw, layer)\u001b[39m\n\u001b[32m     43\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m adata.raw.X\n\u001b[32m     44\u001b[39m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m---> \u001b[39m\u001b[32m45\u001b[39m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43madata\u001b[49m\u001b[43m.\u001b[49m\u001b[43mX\u001b[49m\n",
      "\u001b[31mAttributeError\u001b[39m: 'function' object has no attribute 'X'"
     ]
    }
   ],
   "source": [
    "aa = adata.copy()\n",
    "sc.pp.calculate_qc_metrics(aa, inplace=True)\n",
    "\n",
    "sc.pl.violin(aa, \"n_genes_by_counts\", jitter=0.4, size=1, log=True, groupby=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30f1b325-d724-4ed7-8d77-f6f72c476755",
   "metadata": {},
   "source": [
    "This dataset has relatively low depth, so it is advisable to choose a `max_rank` around `800-1000` (from the default `1500`)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "f9737788-5740-4c5f-b327-fad793ba48c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "uc.compute_ucell_scores(adata, signatures=signatures, max_rank=800)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2a096df-bfc9-4049-8b95-5f3f843f2ca9",
   "metadata": {},
   "source": [
    "This is even more important when applying pyUCell to technologies/modalities of much lower dimensionality, for example probe-based spatial transcriptomics data (e.g. Xenium, CosMx), or antibody tags (ADT) in CITE-seq experiments. Xenium panels contain a few hundred/a few thousand genes; CITE-seq can detect a few hundred proteins, as opposed to thousands of genes in scRNA-seq. The `max_rank` parameter should then also be adjusted to reflect the new dimensionality and set to no more than the number of probes in the panel."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acb3e088-b458-4860-999e-2c7b88ef33fc",
   "metadata": {},
   "source": [
    "## 3. Handling missing genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9650883-263d-4ae7-9760-05ca3455b336",
   "metadata": {},
   "source": [
    "If a subset of the genes in your signature are absent from the count matrix, how should they be handled?\n",
    "\n",
    "pyUCell offers two alternative ways of handling missing genes:\n",
    "- `missing_genes=\"impute\"` (default): it assumes that absence from the count matrix means zero expression. All values for this gene are imputed to zero.\n",
    "-  `missing_genes=\"skip\"`: simply exclude all missing genes from the signatures; they won't contribute to the scores."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a06dc169-60ac-4ada-858c-9790c68a079b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD8T_UCell</th>\n",
       "      <th>CD4T_UCell</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACAACCAC-1</th>\n",
       "      <td>0.514019</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>781</td>\n",
       "      <td>6.661855</td>\n",
       "      <td>2421.0</td>\n",
       "      <td>7.792349</td>\n",
       "      <td>47.748864</td>\n",
       "      <td>63.279637</td>\n",
       "      <td>74.969021</td>\n",
       "      <td>88.393226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGAGCTAC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1352</td>\n",
       "      <td>7.210080</td>\n",
       "      <td>4903.0</td>\n",
       "      <td>8.497807</td>\n",
       "      <td>45.502753</td>\n",
       "      <td>61.023863</td>\n",
       "      <td>71.813176</td>\n",
       "      <td>82.622884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGATCAGC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035692</td>\n",
       "      <td>1131</td>\n",
       "      <td>7.031741</td>\n",
       "      <td>3149.0</td>\n",
       "      <td>8.055158</td>\n",
       "      <td>41.314703</td>\n",
       "      <td>53.794856</td>\n",
       "      <td>65.449349</td>\n",
       "      <td>79.961893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGCTTCCG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>960</td>\n",
       "      <td>6.867974</td>\n",
       "      <td>2639.0</td>\n",
       "      <td>7.878534</td>\n",
       "      <td>39.029936</td>\n",
       "      <td>52.898825</td>\n",
       "      <td>66.691929</td>\n",
       "      <td>82.569155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGTATGCG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>522</td>\n",
       "      <td>6.259581</td>\n",
       "      <td>981.0</td>\n",
       "      <td>6.889591</td>\n",
       "      <td>44.852192</td>\n",
       "      <td>55.657492</td>\n",
       "      <td>67.176351</td>\n",
       "      <td>97.757390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCGAACTCTCAT-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1155</td>\n",
       "      <td>7.052721</td>\n",
       "      <td>3461.0</td>\n",
       "      <td>8.149602</td>\n",
       "      <td>39.237215</td>\n",
       "      <td>52.528171</td>\n",
       "      <td>65.761341</td>\n",
       "      <td>81.074834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCTACTGAGGCA-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1227</td>\n",
       "      <td>7.113142</td>\n",
       "      <td>3447.0</td>\n",
       "      <td>8.145550</td>\n",
       "      <td>37.278793</td>\n",
       "      <td>51.900203</td>\n",
       "      <td>63.881636</td>\n",
       "      <td>78.909196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCTACTTCCTCG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>622</td>\n",
       "      <td>6.434547</td>\n",
       "      <td>1684.0</td>\n",
       "      <td>7.429521</td>\n",
       "      <td>45.783848</td>\n",
       "      <td>61.638955</td>\n",
       "      <td>74.940618</td>\n",
       "      <td>92.755344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGAGAGGC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>454</td>\n",
       "      <td>6.120297</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>6.932448</td>\n",
       "      <td>48.046875</td>\n",
       "      <td>62.402344</td>\n",
       "      <td>75.195312</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGCCTCAC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>724</td>\n",
       "      <td>6.586172</td>\n",
       "      <td>1985.0</td>\n",
       "      <td>7.593878</td>\n",
       "      <td>46.549118</td>\n",
       "      <td>61.964736</td>\n",
       "      <td>73.602015</td>\n",
       "      <td>88.715365</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2700 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CD8T_UCell  CD4T_UCell  n_genes_by_counts  \\\n",
       "index                                                         \n",
       "AAACATACAACCAC-1    0.514019    0.000000                781   \n",
       "AAACATTGAGCTAC-1    0.000000    0.000000               1352   \n",
       "AAACATTGATCAGC-1    0.000000    0.035692               1131   \n",
       "AAACCGTGCTTCCG-1    0.000000    0.000000                960   \n",
       "AAACCGTGTATGCG-1    0.000000    0.000000                522   \n",
       "...                      ...         ...                ...   \n",
       "TTTCGAACTCTCAT-1    0.000000    0.000000               1155   \n",
       "TTTCTACTGAGGCA-1    0.000000    0.000000               1227   \n",
       "TTTCTACTTCCTCG-1    0.000000    0.000000                622   \n",
       "TTTGCATGAGAGGC-1    0.000000    0.000000                454   \n",
       "TTTGCATGCCTCAC-1    0.000000    0.000000                724   \n",
       "\n",
       "                  log1p_n_genes_by_counts  total_counts  log1p_total_counts  \\\n",
       "index                                                                         \n",
       "AAACATACAACCAC-1                 6.661855        2421.0            7.792349   \n",
       "AAACATTGAGCTAC-1                 7.210080        4903.0            8.497807   \n",
       "AAACATTGATCAGC-1                 7.031741        3149.0            8.055158   \n",
       "AAACCGTGCTTCCG-1                 6.867974        2639.0            7.878534   \n",
       "AAACCGTGTATGCG-1                 6.259581         981.0            6.889591   \n",
       "...                                   ...           ...                 ...   \n",
       "TTTCGAACTCTCAT-1                 7.052721        3461.0            8.149602   \n",
       "TTTCTACTGAGGCA-1                 7.113142        3447.0            8.145550   \n",
       "TTTCTACTTCCTCG-1                 6.434547        1684.0            7.429521   \n",
       "TTTGCATGAGAGGC-1                 6.120297        1024.0            6.932448   \n",
       "TTTGCATGCCTCAC-1                 6.586172        1985.0            7.593878   \n",
       "\n",
       "                  pct_counts_in_top_50_genes  pct_counts_in_top_100_genes  \\\n",
       "index                                                                       \n",
       "AAACATACAACCAC-1                   47.748864                    63.279637   \n",
       "AAACATTGAGCTAC-1                   45.502753                    61.023863   \n",
       "AAACATTGATCAGC-1                   41.314703                    53.794856   \n",
       "AAACCGTGCTTCCG-1                   39.029936                    52.898825   \n",
       "AAACCGTGTATGCG-1                   44.852192                    55.657492   \n",
       "...                                      ...                          ...   \n",
       "TTTCGAACTCTCAT-1                   39.237215                    52.528171   \n",
       "TTTCTACTGAGGCA-1                   37.278793                    51.900203   \n",
       "TTTCTACTTCCTCG-1                   45.783848                    61.638955   \n",
       "TTTGCATGAGAGGC-1                   48.046875                    62.402344   \n",
       "TTTGCATGCCTCAC-1                   46.549118                    61.964736   \n",
       "\n",
       "                  pct_counts_in_top_200_genes  pct_counts_in_top_500_genes  \n",
       "index                                                                       \n",
       "AAACATACAACCAC-1                    74.969021                    88.393226  \n",
       "AAACATTGAGCTAC-1                    71.813176                    82.622884  \n",
       "AAACATTGATCAGC-1                    65.449349                    79.961893  \n",
       "AAACCGTGCTTCCG-1                    66.691929                    82.569155  \n",
       "AAACCGTGTATGCG-1                    67.176351                    97.757390  \n",
       "...                                       ...                          ...  \n",
       "TTTCGAACTCTCAT-1                    65.761341                    81.074834  \n",
       "TTTCTACTGAGGCA-1                    63.881636                    78.909196  \n",
       "TTTCTACTTCCTCG-1                    74.940618                    92.755344  \n",
       "TTTGCATGAGAGGC-1                    75.195312                   100.000000  \n",
       "TTTGCATGCCTCAC-1                    73.602015                    88.715365  \n",
       "\n",
       "[2700 rows x 10 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "signatures = {\"CD8T\": [\"CD8A+\", \"CD8B+\", \"CD4-\", \"notagene\"]}\n",
    "uc.compute_ucell_scores(adata, signatures=signatures, missing_genes=\"impute\")\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "adf95ea1-fd9f-47e3-acee-4173e122cfeb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CD8T_UCell</th>\n",
       "      <th>CD4T_UCell</th>\n",
       "      <th>n_genes_by_counts</th>\n",
       "      <th>log1p_n_genes_by_counts</th>\n",
       "      <th>total_counts</th>\n",
       "      <th>log1p_total_counts</th>\n",
       "      <th>pct_counts_in_top_50_genes</th>\n",
       "      <th>pct_counts_in_top_100_genes</th>\n",
       "      <th>pct_counts_in_top_200_genes</th>\n",
       "      <th>pct_counts_in_top_500_genes</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACATACAACCAC-1</th>\n",
       "      <td>0.770771</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>781</td>\n",
       "      <td>6.661855</td>\n",
       "      <td>2421.0</td>\n",
       "      <td>7.792349</td>\n",
       "      <td>47.748864</td>\n",
       "      <td>63.279637</td>\n",
       "      <td>74.969021</td>\n",
       "      <td>88.393226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGAGCTAC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1352</td>\n",
       "      <td>7.210080</td>\n",
       "      <td>4903.0</td>\n",
       "      <td>8.497807</td>\n",
       "      <td>45.502753</td>\n",
       "      <td>61.023863</td>\n",
       "      <td>71.813176</td>\n",
       "      <td>82.622884</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACATTGATCAGC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.035692</td>\n",
       "      <td>1131</td>\n",
       "      <td>7.031741</td>\n",
       "      <td>3149.0</td>\n",
       "      <td>8.055158</td>\n",
       "      <td>41.314703</td>\n",
       "      <td>53.794856</td>\n",
       "      <td>65.449349</td>\n",
       "      <td>79.961893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGCTTCCG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>960</td>\n",
       "      <td>6.867974</td>\n",
       "      <td>2639.0</td>\n",
       "      <td>7.878534</td>\n",
       "      <td>39.029936</td>\n",
       "      <td>52.898825</td>\n",
       "      <td>66.691929</td>\n",
       "      <td>82.569155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCGTGTATGCG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>522</td>\n",
       "      <td>6.259581</td>\n",
       "      <td>981.0</td>\n",
       "      <td>6.889591</td>\n",
       "      <td>44.852192</td>\n",
       "      <td>55.657492</td>\n",
       "      <td>67.176351</td>\n",
       "      <td>97.757390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCGAACTCTCAT-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1155</td>\n",
       "      <td>7.052721</td>\n",
       "      <td>3461.0</td>\n",
       "      <td>8.149602</td>\n",
       "      <td>39.237215</td>\n",
       "      <td>52.528171</td>\n",
       "      <td>65.761341</td>\n",
       "      <td>81.074834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCTACTGAGGCA-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1227</td>\n",
       "      <td>7.113142</td>\n",
       "      <td>3447.0</td>\n",
       "      <td>8.145550</td>\n",
       "      <td>37.278793</td>\n",
       "      <td>51.900203</td>\n",
       "      <td>63.881636</td>\n",
       "      <td>78.909196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTCTACTTCCTCG-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>622</td>\n",
       "      <td>6.434547</td>\n",
       "      <td>1684.0</td>\n",
       "      <td>7.429521</td>\n",
       "      <td>45.783848</td>\n",
       "      <td>61.638955</td>\n",
       "      <td>74.940618</td>\n",
       "      <td>92.755344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGAGAGGC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>454</td>\n",
       "      <td>6.120297</td>\n",
       "      <td>1024.0</td>\n",
       "      <td>6.932448</td>\n",
       "      <td>48.046875</td>\n",
       "      <td>62.402344</td>\n",
       "      <td>75.195312</td>\n",
       "      <td>100.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TTTGCATGCCTCAC-1</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>724</td>\n",
       "      <td>6.586172</td>\n",
       "      <td>1985.0</td>\n",
       "      <td>7.593878</td>\n",
       "      <td>46.549118</td>\n",
       "      <td>61.964736</td>\n",
       "      <td>73.602015</td>\n",
       "      <td>88.715365</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2700 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  CD8T_UCell  CD4T_UCell  n_genes_by_counts  \\\n",
       "index                                                         \n",
       "AAACATACAACCAC-1    0.770771    0.000000                781   \n",
       "AAACATTGAGCTAC-1    0.000000    0.000000               1352   \n",
       "AAACATTGATCAGC-1    0.000000    0.035692               1131   \n",
       "AAACCGTGCTTCCG-1    0.000000    0.000000                960   \n",
       "AAACCGTGTATGCG-1    0.000000    0.000000                522   \n",
       "...                      ...         ...                ...   \n",
       "TTTCGAACTCTCAT-1    0.000000    0.000000               1155   \n",
       "TTTCTACTGAGGCA-1    0.000000    0.000000               1227   \n",
       "TTTCTACTTCCTCG-1    0.000000    0.000000                622   \n",
       "TTTGCATGAGAGGC-1    0.000000    0.000000                454   \n",
       "TTTGCATGCCTCAC-1    0.000000    0.000000                724   \n",
       "\n",
       "                  log1p_n_genes_by_counts  total_counts  log1p_total_counts  \\\n",
       "index                                                                         \n",
       "AAACATACAACCAC-1                 6.661855        2421.0            7.792349   \n",
       "AAACATTGAGCTAC-1                 7.210080        4903.0            8.497807   \n",
       "AAACATTGATCAGC-1                 7.031741        3149.0            8.055158   \n",
       "AAACCGTGCTTCCG-1                 6.867974        2639.0            7.878534   \n",
       "AAACCGTGTATGCG-1                 6.259581         981.0            6.889591   \n",
       "...                                   ...           ...                 ...   \n",
       "TTTCGAACTCTCAT-1                 7.052721        3461.0            8.149602   \n",
       "TTTCTACTGAGGCA-1                 7.113142        3447.0            8.145550   \n",
       "TTTCTACTTCCTCG-1                 6.434547        1684.0            7.429521   \n",
       "TTTGCATGAGAGGC-1                 6.120297        1024.0            6.932448   \n",
       "TTTGCATGCCTCAC-1                 6.586172        1985.0            7.593878   \n",
       "\n",
       "                  pct_counts_in_top_50_genes  pct_counts_in_top_100_genes  \\\n",
       "index                                                                       \n",
       "AAACATACAACCAC-1                   47.748864                    63.279637   \n",
       "AAACATTGAGCTAC-1                   45.502753                    61.023863   \n",
       "AAACATTGATCAGC-1                   41.314703                    53.794856   \n",
       "AAACCGTGCTTCCG-1                   39.029936                    52.898825   \n",
       "AAACCGTGTATGCG-1                   44.852192                    55.657492   \n",
       "...                                      ...                          ...   \n",
       "TTTCGAACTCTCAT-1                   39.237215                    52.528171   \n",
       "TTTCTACTGAGGCA-1                   37.278793                    51.900203   \n",
       "TTTCTACTTCCTCG-1                   45.783848                    61.638955   \n",
       "TTTGCATGAGAGGC-1                   48.046875                    62.402344   \n",
       "TTTGCATGCCTCAC-1                   46.549118                    61.964736   \n",
       "\n",
       "                  pct_counts_in_top_200_genes  pct_counts_in_top_500_genes  \n",
       "index                                                                       \n",
       "AAACATACAACCAC-1                    74.969021                    88.393226  \n",
       "AAACATTGAGCTAC-1                    71.813176                    82.622884  \n",
       "AAACATTGATCAGC-1                    65.449349                    79.961893  \n",
       "AAACCGTGCTTCCG-1                    66.691929                    82.569155  \n",
       "AAACCGTGTATGCG-1                    67.176351                    97.757390  \n",
       "...                                       ...                          ...  \n",
       "TTTCGAACTCTCAT-1                    65.761341                    81.074834  \n",
       "TTTCTACTGAGGCA-1                    63.881636                    78.909196  \n",
       "TTTCTACTTCCTCG-1                    74.940618                    92.755344  \n",
       "TTTGCATGAGAGGC-1                    75.195312                   100.000000  \n",
       "TTTGCATGCCTCAC-1                    73.602015                    88.715365  \n",
       "\n",
       "[2700 rows x 10 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uc.compute_ucell_scores(adata, signatures=signatures, missing_genes=\"skip\")\n",
    "adata.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fe418cb-1f42-4cdd-b5db-ef334e60f640",
   "metadata": {},
   "source": [
    "## 4. Chunk size"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5fd92bef-6be6-4eda-bc23-01eae048b9ae",
   "metadata": {},
   "source": [
    "UCell scores are calculated individually for each cell (though they may be later [smoothed](https://pyucell.readthedocs.io/en/latest/notebooks/knn.html) by nearest-neighbor similarity). This means that computation can be easily split into batches, reducing the computational footprint of gene ranking and enabling parallel processing (see below). The size of the batches is controlled by the `chunk_size` parameter. Large chunks take up more RAM, while small chunk sizes have large overhead from dataset splitting and merging. A sweet spot for `chunk_size` is usually in the order of 100-1000 cells per batch."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b6b0cacc-f978-4ce7-b275-f322be7c51ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "uc.compute_ucell_scores(adata, signatures=signatures, chunk_size=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "628e40e9-7c40-4c9a-a925-1e924403ce15",
   "metadata": {},
   "source": [
    "## 5. Parallelization"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82de55c6-4287-4478-a43f-a8e6f180e69f",
   "metadata": {},
   "source": [
    "Parallelization is handled internally by `joblib` and the `Parallel` module. You may control the number of jobs with the `n_jobs` parameter. By default all available cores are used (`n_jobs=-1`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "27511875-fdc4-4fc3-815f-1e5dc5292959",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 52.4 ms, sys: 55.8 ms, total: 108 ms\n",
      "Wall time: 3.74 s\n"
     ]
    }
   ],
   "source": [
    "%time uc.compute_ucell_scores(adata, signatures=signatures, n_jobs=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b0003f99-e5d7-41d1-bbbd-7b9d425fd46a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 754 ms, sys: 24.6 ms, total: 779 ms\n",
      "Wall time: 778 ms\n"
     ]
    }
   ],
   "source": [
    "%time uc.compute_ucell_scores(adata, signatures=signatures, n_jobs=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efabd22b-536e-4a78-9a30-330ee3358a77",
   "metadata": {},
   "source": [
    "## References"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ccab6d0a-880b-4590-9d34-88c62bf3b50f",
   "metadata": {},
   "source": [
    "If you used UCell in your research, please cite:\n",
    "\n",
    "**UCell: robust and scalable single-cell gene signature scoring.** Massimo Andreatta & Santiago J Carmona (2021) \n",
    "CSBJ - [doi.org/10.1016/j.csbj.2021.06.043](https://doi.org/10.1016/j.csbj.2021.06.043)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68dd321f-9c1f-4247-81ab-343c50968382",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
